### PR TITLE
[VMware/NGC] VMware vSphere OpenSDS Storage Plugin

### PR DESCRIPTION
**Pull request by himanshuvar**
**What this PR does / why we need it**:
With reference to opensds/design-specs#38, This PR is to
support VMware vSphere web client plugin  for OpenSDS storage and third party storage. 

**Special notes for your reviewer**: 
This is a work in progress PR, which includes

1. NGC Plugin: Which is the module to communicate with vSphere components.
2. Adapter Manager : This is the adapter implementation of opensds and thirdparty storage.

compilation guide can be seen in /vmware/ngc/doc
known limitations:
1. text  formatting of some files.
2. opensds adapter for taking default(first) network adapter for picking Initiator IP.
3. default block profile should be created for OpenSDS storage.

**Testing notes**:
1. Tested OpenSDS storage platform addition.
2. Tested OpenSDS login for OpenSDS platform.
3. Tested stoarge pool listing.
4. Tested Volume Creation.
5. Tested Volume Attachment to the ESXI host.

**Release note**:

### CODE REVIEW

**Comment by NajmudheenCT** [comment]
@littlehotcx @skdwriting @wisererik @shanu04 Please review.

**Comment by NajmudheenCT** [comment]
I have tested this PR by integrating changes in #251  .
Steps to test.
1. Clone PR #252 (C:/nbp)
2. clone PR #251 . Copy NGC_Register folder to  c:/nbp/vmware\ngc
3. Run ngc_package.bat (this should create a zip file (/release/NGC-Register-1.0.0)
4.unzip NGC-Register-1.0.0
5.run \NGC-Register-1.0.0\bin\run.bat
6. launch https://localhost:8088/homePage
7. Give vcenter credential and register.
8 login to vcenter server appliance.
9. stop and atart vsphere-client service ( service-control --stop vsphere-client
service-control --start vsphere-client)
10. launch vsphere web client.
11. OpenSDS Storage plugin should be listed in Home page.
12 use plugin to register opensds.
13 provison volume from opensds using volume.

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateDatastoreTask.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @Override
+    public void setContext(Map context) {
+        this.volumeMO = (VolumeMO)context.get(VolumeMO.class.getName());
+    }
+
+    @Override
+    public void runTask() throws Exception {
+        logger.info("---------Step three, running create the datastoreInfo task and convert volume to Datastore...");
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateDatastoreTask.java]: step three ?? this order is applicable only when we follow 
1. create volume
2. attach
3. create data store.

We might have to call this on attached volume .  in that case logging step three is not logical

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Override
+    public void setContext(Map context) {
+        this.context = context;
+    }
+
+    @Override
+    public void runTask() throws Exception {
+
+        logger.info("---------Step one, run create the datastoreInfo task and create volume...");
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java]: same as above "step one" can be removed

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        }
+    }
+
+    @Override
+    public void rollBack() throws Exception {
+        if (volumeMO == null) {
+            logger.info("Could not create the volume...");
+            return;
+        }
+        logger.info("---------Step one, rolling back and deleted created volume...");
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java]: "step one" can be deleted

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+
+    @Override
+    public void setContext(Map context) {
+        this.volumeMO = (VolumeMO)context.get(VolumeMO.class.getName());
+    }
+
+    @Override
+    public void runTask() throws Exception {
+
+        logger.info("---------Step two, run create the datastoreInfo task and mount volume...");
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java]: "step two" can be deleted

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+            changeTaskState(taskInfoList, TaskInfoConst.Status.ERROR,
+                    String.format("Mount volume failed: %s", e.getMessage()));
+            throw e;
+        }
+    }
+
+    @Override
+    public void rollBack() throws Exception {
+        if (connectMO == null) {
+            logger.info("---------Step two, do not need roll back for unmount volume.");
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java]: same here

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+    private static final int HUNDRED = 100;
+
+    private static final int THREE = 3;
+
+    private MathUtil() {
+
+    }
+
+    /**
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java]: delete special charecter from comment

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        BigDecimal newBig = bigDec.divide(one,
+                precision,
+                BigDecimal.ROUND_HALF_UP);
+        return newBig.doubleValue();
+    }
+
+    /**
+     * 向下转换
+     *
+     * @param number 数值
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java]: same here

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+        {
+            max = precision[0];
+        }
+        else if (precision.length == PRECISION_TWO)
+        {
+            max = precision[0];
+            min = precision[1];
+        }
+        NumberFormat nf = NumberFormat.getNumberInstance();
+        //添加舍入模式
```
Comment (NajmudheenCT) [inline vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java]: delete this comment

File: vmware/ngc/doc/OpenSDS_NGC_Compilation_Guide.md [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+Step 2 Extract the downloaded vSphere Client SDK to a local directory.
+```
+![vSphere Client SDK](vSphere_SDK.PNG)
+
+```
+Step 3 Set the system environment variable VSPHERE_SDK_HOME. The variable value is the directory of the
+decompressed SDK package.
+```
+```
+Step 4 In order to support vSphere Web Client Client (Flex ), please also download the vSphere Clint SDK
```
Comment (NajmudheenCT) [inline vmware/ngc/doc/OpenSDS_NGC_Compilation_Guide.md]: Clint -> Client

File: vmware/ngc/NGC-AdapterManager/src/main/java/org/opensds/vmware/ngc/adapters/opensds/RestClient.java [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+             request.close();
+        } catch (Exception e) {
+            // Ignore any exception here
+        } finally {
+            request = null;
+        }
+    }
+
+    void findDeviceInfo(String ip, int port) throws Exception {
+	    request.setUrl(String.format("http://%s:%d/v1beta", ip, port));
```
[the remark this answers is not in the document]
Comment (himanshuvar) [reply vmware/ngc/NGC-AdapterManager/src/main/java/org/opensds/vmware/ngc/adapters/opensds/RestClient.java]: We will refactor this based upon the Backend API response. Once the API are stabilized, we can include the information.
http://petstore.swagger.io/?url=https://raw.githubusercontent.com/opensds/opensds/master/openapi-spec/swagger.yaml#/API%20versions/get__apiVersion_

File: vmware/ngc/NGC-AdapterManager/src/main/java/org/opensds/vmware/ngc/adapters/opensds/OpenSDS.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.json.JSONArray;
+import org.json.JSONObject;
+import org.opensds.vmware.ngc.common.Storage;
+import org.opensds.vmware.ngc.models.*;
+
+class VolumeMOBuilder {
```
[the remark this answers is not in the document]
Comment (himanshuvar) [reply vmware/ngc/NGC-AdapterManager/src/main/java/org/opensds/vmware/ngc/adapters/opensds/OpenSDS.java]: We will be raising a PR to add logging and more Error handling

File: vmware/ngc/NGC-AdapterManager/src/main/java/org/opensds/vmware/ngc/adapters/opensds/Constants.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+public enum Constants {
+	OPENSDS_TENANT,
+	OPENSDS_DOMAIN,
+	OPENSDS_AVAILABILITYZONE,
+	OPENSDS_STORAGENAME,
+	OPENSDS_VENDOR;
+
+	private static final String PATH = "/constants.properties";
```
[the remark this answers is not in the document]
Comment (himanshuvar) [reply vmware/ngc/NGC-AdapterManager/src/main/java/org/opensds/vmware/ngc/adapters/opensds/Constants.java]: The idea is to include all the Constants related to the AdapterManager

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateDatastoreTask.java (hunk 1 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateDatastoreTask.java]: Done

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java (hunk 2 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java]: Done

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java (hunk 3 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/CreateLunTask.java]: Done

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java (hunk 4 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java]: Done

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java (hunk 5 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/task/createDatastore/MountLunTask.java]: Done

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java (hunk 6 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java]: Done

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java (hunk 7 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java]: Done

File: vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java (hunk 8 above)
Comment (himanshuvar) [reply vmware/ngc/NGC-Plugin/src/main/java/org/opensds/vmware/ngc/util/MathUtil.java]: Done

File: vmware/ngc/doc/OpenSDS_NGC_Compilation_Guide.md (hunk 9 above)
Comment (himanshuvar) [reply vmware/ngc/doc/OpenSDS_NGC_Compilation_Guide.md]: Done

**Pull request merged**